discovery/tailscale/tailscale.go

```diff
@@ -0,0 +1,225 @@
+	client          *tailscale.Client
+}
+
+func NewDiscovery(conf *SDConfig, logger log.Logger, metrics discovery.DiscovererMetrics) (*Discovery, error) {
+	m, ok := metrics.(*tailscaleMetrics)
+	if !ok {
+		return nil, fmt.Errorf("invalid discovery metrics type")
+	}
+
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	if conf.URL == "" {
+		conf.URL = "https://api.tailscale.com"
+	}
+
+	tsClient := tailscale.NewClient(
+		conf.Tailnet,
+		tailscale.APIKey(conf.APIToken),
+	)
+	tsClient.BaseURL = conf.URL
+
+	d := &Discovery{
+		port:            conf.Port,
+		client:          tsClient,
+		url:             conf.URL,
+		tagSeparator:    conf.TagSeparator,
+		refreshInterval: time.Duration(conf.RefreshInterval),
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		refresh.Options{
+			Logger:              logger,
+			Mech:                "tailscale",
+			Interval:            time.Duration(conf.RefreshInterval),
+			RefreshF:            d.refresh,
+			MetricsInstantiator: m.refreshMetrics,
+		},
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	tsDevices, err := d.client.Devices(ctx, tailscale.DeviceDefaultFields)
+	if err != nil {
+		failuresCount.Inc()
+		return nil, err
+	}
+
+	tg := &targetgroup.Group{
+		// Use a pseudo-URL as source.
+		Source: d.url,
+	}
+
+	for _, device := range tsDevices {
+		labels := model.LabelSet{
+			tsLabelDeviceID:                         model.LabelValue(device.DeviceID),
+			tsLabelDeviceOS:                         model.LabelValue(device.OS),
+			tsLabelDeviceClientVersion:              model.LabelValue(device.ClientVersion),
+			tsLabelDeviceName:                       model.LabelValue(device.Name),
+			tsLabelDeviceHostname:                   model.LabelValue(device.Hostname),
+			tsLabelDeviceUser:                       model.LabelValue(device.User),
+			tsLabelDeviceIsExternal:                 model.LabelValue(strconv.FormatBool(device.IsExternal)),
+			tsLabelDeviceAuthorized:                 model.LabelValue(strconv.FormatBool(device.Authorized)),
+			tsLabelDeviceUpdateAvailable:            model.LabelValue(strconv.FormatBool(device.UpdateAvailable)),
+			tsLabelDeviceKeyExpiryDisabled:          model.LabelValue(strconv.FormatBool(device.KeyExpiryDisabled)),
+			tsLabelDeviceBlocksIncommingConnections: model.LabelValue(strconv.FormatBool(device.BlocksIncomingConnections)),
+		}
+
+		addr := net.JoinHostPort(device.Addresses[0], strconv.FormatUint(uint64(d.port), 10))
+		labels[model.AddressLabel] = model.LabelValue(addr)
+
+		if len(device.Addresses) > 0 {
+			addresses := d.tagSeparator + strings.Join(device.Addresses, d.tagSeparator) + d.tagSeparator
+			labels[tsLabelDeviceAddresses] = model.LabelValue(addresses)
+		}
+
+		if len(device.Tags) > 0 {
+			tags := d.tagSeparator + strings.Join(device.Tags, d.tagSeparator) + d.tagSeparator
+			labels[tsLabelDeviceTags] = model.LabelValue(tags)
+		}
+		tg.Targets = append(tg.Targets, labels)
```

Comment:
Wouldn't you want to have this line at the end of the `for` loop, after setting the other two labels below? Maybe it doesn't make much of a difference because you're still manipulating the same map reference, but still...

Reply:
Thinking this might just be a consequence of some code being written after the initial design. Fixed it in the latest version.